### ISSUE TITLE
Tree depth doesn't match expectations

### ISSUE THREAD

**Issue by the5thbeatle** [14corman/AIDescisionTree#16]
What's happening:
Setting the tree depth to (Some 0) or (Some 1) builds an entire tree.

What's expected:
Setting the tree depth to (Some 0) guesses the majority label for all examples.
Setting the tree depth to (Some 1) produces a tree with at most one split.

Steps to reproduce:
Call cross_validate with a dataset, setting d=(Some 0) or d=(Some 1).
Print the resulting graph. There should be many more nodes than expected.

**Comment by 14corman** [comment]
0 as input should fail. You cannot have a tree of depth = 0 since then you do not have a tree. That is probably why the book starts at 1. For depth = 1, I will look into that and see why it may be building the whole tree. Are you sure you got the right code? I know I fixed depth a few patches ago. You may have an older version of the code.

**Comment by the5thbeatle** [comment]
0 shouldn't fail. A depth of 0 means you only have the root node, which
should be a single leaf node that guesses the majority class label.

The code is up-to-date.

**Comment by 14corman** [comment]
No depth of 0 means nothing. A depth of 1 means you have a root.

**Comment by the5thbeatle** [comment]
The depth of a node is equal to the number of edges on the path from it to
the root. For the root, that is a depth of 0. For the first set of of
children, the depth is 1.

**Comment by 14corman** [comment]
The depth of the node is equal to the height of the node tree at that node. A root node is at height 1 so it has a depth of 1.

**Comment by the5thbeatle** [comment]
The height of one-node tree is 0. Height is equal to the number of edges
from the root to the deepest node in the tree.

From the Wikipedia on trees:
"The root node has depth zero, leaf nodes have height zero, and a tree with
only a single node (hence both a root and leaf) has depth and height zero."

**Comment by 14corman** [comment]
Ok I fixed it so depth should work. It started at depth = 1 and checked if depth + 1 = max depth. This means that if max depth was 0 or 1 then it would never equal and it would build the whole tree. I set it to start at -1 so that it will check for 0.

**Comment by 14corman** [comment]
I wanted to reopen this just so people can see it and make sure it is right. Here is the graph for depth 0:
![graphviz (11)](https://user-images.githubusercontent.com/15676385/57170204-dffc7480-6dd0-11e9-94b4-8eba79d35444.png)

This graph has an F1 score of 0.46.

The graph with depth 1:
![graphviz (12)](https://user-images.githubusercontent.com/15676385/57170222-f86c8f00-6dd0-11e9-9c2c-e3102638198c.png)

This graph has an F1 score of 0.96.

There was a bit more to the Decision Tree code that I had to change than I thought. I not only had to start at -1 for depth, but I also had to check if a node was at the max depth before it went to do any calculations.
